content/reference/v1beta1/repeatable.md

```diff
@@ -0,0 +1,278 @@
+```yaml
+    - port: '{{repl ConfigOption "[[repl .service_ports ]]" | ParseInt }}'
+```
+This repeatable templating is not compatible with sprig templating functions.  It is designed for inserting repeatable `keys` into the manifest. Repeatable templating **can** be placed inside of Replicated config templating.
+
+## Ordering
+
+Repeatable templates are processed before config template rendering.
+
+Repeatable items are processed in order of the template targets in the Config Spec file.  Effectively, this ordering is from the top of the Config Spec, by Config Group, by Config Item, and then by template target.
+
+```yaml
+    - name: ports
+      items:
+      - name: service_ports
+        title: Service Ports
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: v1 #processed first
+          kind: Service
+          name: my-service
+          namespace: my-app
+          yamlPath: 'spec.ports[0]'
+        - apiVersion: v1 #processed second
+          kind: Service
+          name: my-service
+          namespace: my-app
+          yamlPath:
+    {...}
+      - name: other_ports
+        title: Other Service Ports
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: v1 #processed third
+          kind: Service
+          name: my-other-service
+          namespace: my-app
+    {...}
+    - name: deployments
+      items:
+      - name: deployment-name
+        title: Deployment Names
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: apps/v1 #processed fourth
+          kind: Deployment
+          name: my-deployment
+          namespace: my-app
+          yamlPath:
+    {...}
+```
+
+## Repeatable Item Example for a YamlPath
+**Config spec**
+```yaml
+    - name: ports
+      items:
+      - name: service_ports
+        title: Service Ports
+        type: text
+        repeatable: true
+        template:
+        - apiVersion: v1
+          kind: Service
+          name: my-service
+          namespace: my-app
+          yamlPath: 'spec.ports[0]'
+        valuesByGroup:
+          ports:
+            port-default-1: "80"
+            port-default-2: "443"
+```
+**Template manifest**
+```yaml
+    apiVersion: v1
+    kind: Service
+    metadata:
+      name: my-service
+      namespace: my-app
+    spec:
+      type: NodePort
+      ports:
+      - port: '{{repl ConfigOption "[[repl .service_ports ]]" | ParseInt }}'
+        name: '{{repl ConfigOptionName "[[repl .service_ports ]]" }}'
+      selector:
+        app: repeat_example
+        component: my-deployment
+```
+
+**After repeatable config processing**
+```yaml
+    apiVersion: v1
+    kind: Service
+    metadata:
+      name: my-service
+      namespace: my-app
+    spec:
+      type: NodePort
+      ports:
+      - port: '{{repl ConfigOption "port-default-1" | ParseInt }}'
+        name: '{{repl ConfigOptionName "port-default-1" }}'
```

Comment:
Never.  The purpose of the function is for the previous step, `{{repl ConfigOptionName "[[repl .service_ports ]]" }}` to work with generated variable names.  The ConfigOptionName wouldn't be known until the customer's config values are provided.

This part of the example is meant to explain what's happening during the multi-step render process.  If that's unclear, I can provide a short explanation.

Reply:
I added a note at the top of the yaml example here to explain that this section is all internal to KOTS rendering.  Let me know if I've made it more confusing 😄

Reply:
I think I'm at least halfway tracking :) What's confusing to me is that the incremental steps make it look like `ConfigOptionName` just returns whatever it is passed, in which case it doesn't feel necessary. So what I don't totally see yet is the relationship between `ConfigOptionName` and `"[[repl .service_ports]]"`. It sounds like they need to work together, but the incremental thing is throwing me off since it looks like `"[[repl .service_ports]]"` is doing all the work.

Reply:
... somewhere in the dev process this totally became possible and I just missed it.  Originally it required to be inside the repl{{ }} templating.  The following works according to the unit tests.

```yaml
- secret:
     name: 'repl[[ .secretName ]]'
     pod: repl{{ ConfigOption "podName" }}
```

I will update the docs to show this as the primary example instead of using ConfigOptionName.  Thanks for pointing this out!!

Reply:
Glad we stumbled upon that together. So what does `ConfigOptionName` get used for then? I don't think it's used in these examples now in light of that change.

Also, do we need to doc the `'repl[[ .secretName ]]'` syntax since we haven't used that before? It seems different than our other `repl` templating because it has the dot.

Reply:
`repl[[ .variable ]]` is documented under the Templating section here.  It's only used for repeatable items so I think this is the right place.

I've removed the ConfigOptionName documentation as it's unnecessary.

Reply:
Cool. You can resolve this one.